src/core/AutoRest.Core/Validation/RecursiveObjectValidator.cs

```diff
@@ -1,85 +1,77 @@
+    ///     A validator that traverses an object graph, applies validation rules, and logs validation messages
     /// </summary>
     public class RecursiveObjectValidator
     {
-        public RecursiveObjectValidator()
-        {
-        }
-
         public IEnumerable<ValidationMessage> GetValidationExceptions(object entity)
         {
-            return RecursiveValidate(entity, new List<RuleAttribute>());
+            return RecursiveValidate(entity, Enumerable.Empty<Rule>());
         }
 
-        private IEnumerable<ValidationMessage> RecursiveValidate(object entity, IEnumerable<RuleAttribute> collectionRules)
+        private IEnumerable<ValidationMessage> RecursiveValidate(object entity, IEnumerable<Rule> rules)
         {
-            if (entity != null)
+            if (entity == null)
             {
-                if (entity is IList)
-                {
-                    // Recursively validate each list item and add the item index to the location of each validation message
-                    IList<dynamic> list = ((IList)entity).Cast<dynamic>().ToList();
-                    if (list != null)
-                    {
-                        for (int i = 0; i < list.Count; i++)
-                        {
-                            foreach (ValidationMessage exception in RecursiveValidate(list[i], collectionRules))
-                            {
-                                exception.Path.Add($"[{i}]");
-                                yield return exception;
-                            }
-                        }
-                    }
-                }
-                else if (entity is IDictionary)
-                {
-                    // Recursively validate each dictionary entry and add the entry key to the location of each validation message
-                    IDictionary<string, dynamic> dict = ((IDictionary)entity).Cast<dynamic>().ToDictionary(entry => (string)entry.Key, entry => entry.Value);
-                    if (dict != null && entity.IsValidatableDictionary())
-                    {
-                        foreach (var pair in dict)
-                        {
-                            foreach (ValidationMessage exception in RecursiveValidate(pair.Value, collectionRules))
-                            {
-                                exception.Path.Add(pair.Key);
-                                yield return exception;
-                            }
-                        }
-                    }
-                }
-                else if (entity.GetType().IsClass && entity.GetType() != typeof(string))
+                return Enumerable.Empty<ValidationMessage>();
+            }
+
+            // ensure that the rules can be re-enumerated without re-evaluating the enumeration.
+            var collectionRules = rules.ReEnumerable();
+
+            var list = entity as IList;
+            if (list != null)
+            {
+                // Recursively validate each list item and add the 
```

Comment:
Sorry, I couldn't resist rewriting this a bit. 

The original implementation has an extra reallocation of the entire list (rather than being able to `Select` on it directly. 

Same thing with the IDictionary implementation.